File: src/main/java/net/wanhack/model/region/CellMap.java
/*
 *  Copyright 2005 The Wanhack Team
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package net.wanhack.model.region;

import java.io.Serializable;
import java.util.AbstractMap;
import java.util.AbstractSet;
import java.util.Iterator;
import java.util.Map;
import java.util.NoSuchElementException;
import java.util.Set;

import net.wanhack.utils.ObjectUtils;

/**
 * Map from {@link Cell} objects to values.
 */
public final class CellMap<V> 
        extends AbstractMap<Cell, V> implements Serializable {

    private final Region region;
    private final V[] mappings;
    private static final long serialVersionUID = 0;
    
    @SuppressWarnings("unchecked")
    public CellMap(Region region) {
        this.region = region;
        this.mappings = (V[]) new Object[region.getWidth() * region.getHeight()];
    }
    
    @Override
    public V get(Object key) {
        return mappings[index((Cell) key)];
    }
    
    @Override
    public V put(Cell key, V value) {
        int index = index(key);
        V old = mappings[index];
        mappings[index] = value;
        return old;
    }
    
    @Override
    public V remove(Object key) {
        if (key instanceof Cell) {
            return put((Cell) key, null);
        } else {
            return null;
        }
    }
    
    @Override
    public boolean isEmpty() {
        return size() != 0;
    }
    
    @Override
    public int size() {
        int size = 0;
        for (int i = 0; i < mappings.length; i++) {
            if (mappings[i] != null) {
                size++;
            }
        }
        return size;
    }

    @Override
    public void clear() {
        for (int i = 0; i < mappings.length; i++) {
            mappings[i] = null;
        }
    }
    
    @Override
    public Set<Entry<Cell, V>> entrySet() {
        return new EntrySet();
    }
    
    private int index(Cell cell) {
        return cell.x + cell.y * region.getWidth();
    }
    
    private Cell cell(int index) {
        int y = index / region.getWidth();
        int x = index % region.getWidth();
        
        return region.getCell(x, y);
    }
    
    private class EntrySet extends AbstractSet<Entry<Cell, V>> {
        @Override
        public Iterator<Entry<Cell, V>> iterator() {
            return new EntryIterator();
        }
        
        @Override
        public boolean add(Entry<Cell, V> o) {
            put(o.getKey(), o.getValue());
            return true;
        }
        
        @Override
        public boolean remove(Object o) {
            if (o instanceof Entry) {
                Entry entry = (Entry) o;
                
                Object value = get(entry.getKey());
                if (ObjectUtils.equals(value, entry.getValue())) {
                    Object old = CellMap.this.remove(entry.getKey());
                    return old != null;
                } else {
                    return false;
                }
            }
            return false;
        }
        
        @Override
        public boolean contains(Object o) {
            if (o instanceof Entry) {
                Entry entry = (Entry) o;
                
                Object value = get(entry.getKey());
                return ObjectUtils.equals(value, entry.getValue());
            }
            return false;
        }
        
        @Override
        public int size() {
            return CellMap.this.size();
        }
    }
    
    private class CellMapEntry implements Map.Entry<Cell, V> {
        private final int index;

        public CellMapEntry(int index) {
            this.index = index;
        }

        public Cell getKey() {
            return cell(index);
        }
        
        public V getValue() {
            return mappings[index];
        }
        
        public V setValue(V value) {
            V old = mappings[index];
            mappings[index] = value;
            return old;
        }
        
        @Override
        public boolean equals(Object obj) {
            if (obj == this) return true;
            
            if (obj instanceof Map.Entry) {
                Map.Entry rhs = (Map.Entry) obj;
                
                return ObjectUtils.equals(getKey(), rhs.getKey())
                    && ObjectUtils.equals(getValue(), rhs.getValue());
            }

            return false;
        }
        
        @Override
        public int hashCode() {
            Cell key = getKey();
            V value = getValue();
            return key.hashCode() ^ (value == null ? 0 : value.hashCode());
        }
    }
    
    private class EntryIterator implements Iterator<Entry<Cell, V>> {
        private int index = 0;
        private int previous = -1;
        
        public boolean hasNext() {
            while (index < mappings.length && mappings[index] == null) {
                index++;
            }
            
            return index < mappings.length;
        }
        
        public Entry<Cell, V> next() {
            if (hasNext()) {
                previous = index++;
                return new CellMapEntry(previous);
            } else {
                throw new NoSuchElementException();
            }
        }
        
        public void remove() {
            if (previous != -1) {
                mappings[previous] = null;                            
            } else {
                throw new IllegalStateException();
            }
        }
    }
}

File: core/src/main/kotlin/net/wanhack/model/region/CellMap.kt
/*
 * Copyright 2013 The Wanhack Team
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.wanhack.model.region

import java.util.AbstractMap
import java.util.AbstractSet
import java.util.NoSuchElementException
import java.util.Objects

@Suppress("UNCHECKED_CAST")
class CellMap<V : Any>(private val region: Region): AbstractMap<Cell, V>() {

    private val mappings = Array<Any?>(region.width * region.height) { null }

    override fun get(key: Cell?): V? =
        mappings[index((key as Cell).coordinate)] as V?

    override fun put(key: Cell, value: V): V? {
        val index = index(key.coordinate)
        val old = mappings[index]
        mappings[index] = value
        return old as V?
    }

    override fun remove(key: Cell?): V? =
        if (key is Cell) {
            val index = index(key.coordinate)
            val old = mappings[index]
            mappings[index] = null
            old as V?
        } else
            null

    override fun isEmpty() = size != 0

    override val size: Int
        get() = mappings.count { it != null }

    override fun clear() {
        for (i in mappings.indices)
            mappings[i] = null
    }

    override val entries: MutableSet<MutableMap.MutableEntry<Cell, V>>
        get() = EntrySet(this)

    private fun index(c: Coordinate): Int =
        c.x + c.y * region.width

    private fun cell(index: Int): Cell =
        region[index % region.width, index / region.width]

    private class EntrySet<V : Any>(val map: CellMap<V>): AbstractSet<MutableMap.MutableEntry<Cell, V>>() {

        override fun iterator() =
            EntryIterator(map)

        override fun add(element: MutableMap.MutableEntry<Cell, V>): Boolean {
            map[element.key] = element.value
            return true
        }

        override fun remove(element: MutableMap.MutableEntry<Cell, V>?): Boolean {
            if (element is Map.Entry<Any?, Any?>) {
                val key = element.key
                val value = map[key]
                if (value == element.value)
                    return map.remove(key) != null
                else
                    return false
            }

            return false
        }

        override fun contains(element: MutableMap.MutableEntry<Cell, V>?): Boolean =
            element != null && map[element.key] == element.value

        override val size: Int
            get() = map.size
    }

    private class CellMapEntry<V : Any>(val map: CellMap<V>, val index: Int): MutableMap.MutableEntry<Cell, V> {

        override val key: Cell
            get() = map.cell(index)

        override val value: V
            get() = map.mappings[index]!! as V

        override fun setValue(newValue: V): V {
            val old = map.mappings[index]
            map.mappings[index] = newValue
            return old!! as V
        }

        override fun equals(other: Any?): Boolean {
            if (other == this)
                return true

            return other is Map.Entry<Any?, Any?> && key == other.key && value == other.value
        }

        override fun hashCode() = Objects.hash(key, value)
    }

    private class EntryIterator<V : Any>(val map: CellMap<V>): MutableIterator<MutableMap.MutableEntry<Cell, V>> {
        private var index = 0
        private var previous = -1

        override fun hasNext(): Boolean {
            while (index < map.mappings.size && map.mappings[index] == null)
                index++

            return index < map.mappings.size
        }

        override fun next(): MutableMap.MutableEntry<Cell, V> {
            if (hasNext()) {
                previous = index++
                return CellMapEntry(map, previous)
            }
            else
                throw NoSuchElementException()
        }

        override fun remove() {
            if (previous == -1)
                throw IllegalStateException()
            map.mappings[previous] = null
        }
    }
}
